src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1268,4 +1274,101 @@ ss::future<bool> tx_gateway_frontend::try_create_tx_topic() {
       });
 }
 
+void tx_gateway_frontend::expire_old_txs() {
+    (void)ss::with_gate(_gate, [this] {
+        auto shard = _shard_table.local().shard_for(model::tx_manager_ntp);
+
+        if (shard == std::nullopt) {
+            rearm_expire_timer();
+            return ss::now();
+        }
+
+        return container()
+          .invoke_on(
+            *shard,
+            _ssg,
+            [](tx_gateway_frontend& self) {
+                auto partition = self._partition_manager.local().get(
+                  model::tx_manager_ntp);
+                if (!partition) {
+                    vlog(
+                      clusterlog.warn,
+                      "can't get partition by {} ntp",
+                      model::tx_manager_ntp);
+                    return ss::now();
+                }
+
+                auto stm = partition->tm_stm();
+
+                auto txs = stm->get_expired_txs();
+                return self.expire_old_txs(stm, txs);
+            })
+          .finally([this] { rearm_expire_timer(); });
+    });
+}
+
+ss::future<> tx_gateway_frontend::expire_old_txs(
+  ss::shared_ptr<tm_stm> stm, std::vector<kafka::transactional_id> tx_ids) {
+    for (auto tx_id : tx_ids) {
+        co_await expire_old_tx(stm, tx_id);
+    }
+}
+
+ss::future<> tx_gateway_frontend::expire_old_tx(
+  ss::shared_ptr<tm_stm> stm, kafka::transactional_id tx_id) {
+    return stm->get_tx_lock(tx_id)->with([this, stm, tx_id]() {
+        return do_expire_old_tx(
+          stm, tx_id, config::shard_local_cfg().create_topic_timeout_ms());
+    });
+}
+
+ss::future<> tx_gateway_frontend::do_expire_old_tx(
+  ss::shared_ptr<tm_stm> stm,
+  kafka::transactional_id tx_id,
+  model::timeout_clock::duration timeout) {
+    auto tx_opt = co_await stm->get_actual_tx(tx_id);
+    if (!tx_opt) {
+        // either timeout or already expired
+        co_return;
+    }
+
+    auto tx = tx_opt.value();
+    if (!stm->is_expired(tx)) {
+        co_return;
+    }
+
+    checked<tm_transaction, tx_errc> r(tx);
+
+    if (tx.status == tm_transaction::tx_status::ready) {
```

Comment:
The control flow here is a little awkward.

I think `tx_errc ec` isn't required, since `r` can accept the error.

Then I think it could be flattened and turned into a switch to help catch missing states at compile time rather than runtime `vassert`.

Reply:
`r` stores either a transcation or an error code. If we use `r` to store the `ec` code returned by `recommit_tm_tx` or `reabort_tm_tx` we'll lose the transaction object. This is a reason why I introduced `tx_errc ec` variable and overwrite `r` only if `ec` isn't successful.

Reply:
Sorry, I still don't see it. It looks like the value or error of `r` is never used. `r` is only used to detect whether it has a value or an error.

Reply:
yeh I think that `r` is being used as a convenient container to transfer state across all the conditionals. I do agree it could be more explicit, but it looks like it should work correctly.

@rystsov related to Ben's questions bout tx_errc, it does seem like `result<tm_transaction>` is sufficient (or should be assuming that `tx_errc` is defined correct). how is `checked<tm_transaction, tx_errc>` different?

Reply:
I prefer to keep it. it's true that it's possible to flat this out and assign the `reabort / recommit` result to `r` but then we switch the success path from `r` has a tx value to `r` has a tx value or its error is none (the result of the succesful `re*` execution) and it's awkward.